Migrate test and hadoop-mapreduce modules to JUnit5

DomGarguilo opened the pull request:
This PR coverts the accumulo-test and accumulo-hadoop-mapreduce modules from JUnit4 to JUnit5.

These modules had to be bundled together due to some dependency overlap.

TODO:
- ~~Scalable timeouts need to be sorted out ([here is a description of the problem](https://github.com/apache/accumulo/issues/2441#issuecomment-1050166774))~~
  - ~~Remove comments from AccumuloITBase~~
- ~~Finish `@Category` -> `@Tag` conversion~~
  - ~~Ensure they function the same~~

DomGarguilo commented:
I think this is ready for review. As far as I can tell, everything works as before with all tests passing. I know this is a huge PR to review so this might help when looking through:

> Here are some quick notes that might help to reference for those who work on the migration or for reviewers.
> 
> | JUnit4  | JUnit5  |
> | ----------- | ----------- |
> | `@Test` | `@Test` |
> | `@Before` | `@BeforeEach` |
> | `@After` | `@AfterEach` |
> | `@BeforeClass` | `@BeforeAll` |
> | `@AfterClass` | `AfterAll` |
> | `@Ignore` | `@Disable` |
> | `@Category` | `@Tag` |
> 
> | | JUnit4  | JUnit5  |
> | ----------- | ----------- | ----------- |
> | Testing class package | `org.junit` | `org.junit.jupiter.api` |
> | Assertions class | `Assert` | `Assertions` |
> 
> `TemporaryFolder` was removed and instead a `File` or `Path` is used in JUnit5 with the `@TempDir` annotation.
> 
> Additionally, the order of parameters in assert statements has changed. The optional message is now the last parameter rather than the first parameter. e.g. `assertTrue("should be equal", 1==3)` becomes `assertTrue(1==3, "should be equal")`

_Originally posted by @DomGarguilo in https://github.com/apache/accumulo/issues/2441#issuecomment-1041908425_

In addition to the things mentioned above, `@Category` has been replaced with `@Tag` which no longer requires an interface so those were deleted as well. This slightly changes how groups of tests are run. for example to run all tests annotated with the `ZooKeeperTestingServerTests` tag, you would run `mvn clean verify -Dtest=foo -Dfailsafe.groups=ZooKeeperTestingServerTests -Dspotbugs.skip` where as previously you would have to specify the category package i.e., `mvn clean verify -Dtest=foo -Dfailsafe.groups=org.apache.accumulo.test.categories.ZooKeeperTestingServerTests -Dspotbugs.skip`

ctubbsii commented on TESTING.md:
````diff
@@ -98,7 +98,7 @@ the tests.
 
 ```bash
-mvn clean verify -Dtest=foo -Daccumulo.it.properties=/home/user/my_cluster.properties -Dfailsafe.groups=org.apache.accumulo.test.categories.StandaloneCapableClusterTests -Dspotbugs.skip
+mvn clean verify -Dtest=foo -Daccumulo.it.properties=/home/user/my_cluster.properties -Dfailsafe.groups=StandaloneCapableClusterTests -Dspotbugs.skip
````
These tag names can be simplified further. The original names like `SunnyDayTests` was chosen because we had to create a class name that made sense on its own. Now that's not necessary, since the class is deleted, and the fact that we're in a "Test" context is already obvious when using the Tag annotation. So, something like `@Tag("SunnyDay")` should suffice instead of `@Tag("SunnyDayTests")`.

DomGarguilo replied:
Good idea. I'm wondering if in that case we could remove the [`-Psunny` profile](https://github.com/apache/accumulo/blob/97af14509d2755de29e2c0196201018f576cab68/pom.xml#L1548-L1555) since to run the same tests now, it would just be `-Dfailsafe.groups=SunnyDay`. The prior is still shorter but just thought I would point it out and see what you/others thought. I assume the profile was created since this group of tests is run often, and before, one would have to type out the full class name (org.apache.accumulo.test.categories.SunnyDayTests).

ctubbsii replied:
I agree it is superfluous, but it's historical, so it's probably worth keeping. When we initially added these category annotations, I resisted creating a bunch of new profiles with different combinations of activating them, because you can just use the `-Dfailsafe.groups=` property. I still think it's unnecessary to create new profiles, but keeping this one around for historical reasons is probably desirable. If you wanted to simplify it further, you could call it `@Tag("Sunny")` in order to get even closer to the profile name.

DomGarguilo replied:
Addressed in [d1fa1f3](https://github.com/apache/accumulo/pull/2545/commits/d1fa1f31dbbf5e8728ceaa9359f2e92b67cd4d75)

ctubbsii reviewed (changes requested):
Looks good overall. Please don't rebase / force-push this, because I don't want to re-review 274 files :smiley_cat:

I had a few suggestions below.

DomGarguilo commented on test/src/main/java/org/apache/accumulo/test/functional/KerberosRenewalIT.java:
```diff
@@ -137,15 +138,16 @@ public void configureMiniCluster(MiniAccumuloConfigImpl cfg, Configuration coreS
   }
 
   // Intentionally setting the Test annotation timeout. We do not want to scale the timeout.
-  @Test(timeout = TEST_DURATION)
+  @Test
+  @Timeout(value = TEST_DURATION_MINUTES, unit = MINUTES)
```
No, these explicitly set timeouts are not scaled and take precedence over the registered extension.

DomGarguilo commented on hadoop-mapreduce/src/test/java/org/apache/accumulo/WithTestNames.java:
```diff
@@ -16,10 +16,27 @@
  * under the License.
  */
-package org.apache.accumulo.test.categories;
+package org.apache.accumulo;
```
Sounds good. I'll move it into the package you mentioned. Do you think this is something that will also need to happen for [`WithTestNames` in the test module](https://github.com/DomGarguilo/accumulo/blob/testAndHadoopMRmodulesMigrateJUnit/test/src/main/java/org/apache/accumulo/WithTestNames.java)?

DomGarguilo commented on test/src/main/java/org/apache/accumulo/harness/AccumuloClusterHarness.java:
```diff
@@ -60,7 +59,7 @@
  * {@link SharedMiniClusterBase}. This instance can be MAC or a standalone instance.
  */
-@Category(StandaloneCapableClusterTests.class)
+@Tag("StandaloneCapableCluster")
```
I added constants, where possible, to AccumuloITBase.java but for the single `@Tag` that is used outside of accumulo-test module ([here](https://github.com/apache/accumulo/blob/c1507285b4f52785d4c3e4b9fc6ab3ad17b32185/server/monitor/src/test/java/org/apache/accumulo/monitor/it/WebViewsIT.java#L63)), I had to create [a class to store the constant](https://github.com/apache/accumulo/blob/c1507285b4f52785d4c3e4b9fc6ab3ad17b32185/server/monitor/src/test/java/org/apache/accumulo/monitor/TagNameConstants.java). Please let me know what you think of this approach and if you have any alternatives in mind.

ctubbsii replied:
Works for me.

DomGarguilo commented:
@ctubbsii I think I addressed all the comments you had in [c150728](https://github.com/apache/accumulo/pull/2545/commits/c1507285b4f52785d4c3e4b9fc6ab3ad17b32185)

ctubbsii commented on hadoop-mapreduce/src/test/java/org/apache/accumulo/hadoop/WithTestNames.java:
```diff
@@ -16,7 +16,7 @@
  * under the License.
  */
-package org.apache.accumulo;
+package org.apache.accumulo.hadoop;
```
This isn't unique enough to ensure that this package will only appear in the test classes/test jar. This package name is still one that could conceivable collide with a package in the main jar for this module if it is ever referenced by an integration test.

If it's only referenced by unit tests, that's fine, because jar sealing won't come into play... because the maven-surefire-plugin runs before the jars are built. However, integration tests run after the package phase when the jars are built, so the classes used for integration tests must not be in packages that could exist in the main jar.

DomGarguilo replied:
I'm not sure I fully understand but I think I moved them into a package that is only used for tests. The new package for this class reads `org.apache.accumulo.hadoop.its`

ctubbsii commented on server/monitor/src/test/java/org/apache/accumulo/monitor/TagNameConstants.java:
```diff
@@ -0,0 +1,23 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.monitor;
```
Similar comment as above, regarding the uniqueness of the test package name if the class is used for integration tests, due to jar sealing.

ctubbsii reviewed (approved):
I think this is all good now, but I want to run my own full IT build before it is merged, just to be sure.

DomGarguilo commented:
> I think this is all good now, but I want to run my own full IT build before it is merged, just to be sure.

Sounds good. Thanks for your review and feel free to merge this PR after your testing, if you would like to.
